### PR TITLE
BFD-501: Adding x-forwarded and unifying bulk headers to documentation

### PR DESCRIPTION
**Pull request by jzulim**
### Change Details

We are wanting to add the x-forarded-for header to synchronous requests (i.e. BB2.0) and would like to unify our bulk header requests so they are not peer partner specific. 

### Acceptance Validation

Are the header names correct and do they imply their usage correctly?
Do the header descriptions provide the context for future and current users?

### Feedback Requested

Are the header names correct and do they imply their usage correctly?
Do the header descriptions provide the context for future and current users?

### External References

- [BFD-501](https://jira.cms.gov/browse/BFD-507)

### Security Implications

- [ ] new software dependencies

- [ ] altered security controls

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

**Comment by msnook** [comment]
Added @mtrang1263 to reviewer list.

File: docs/request-audit-headers.md [hunk 1]
```diff
@@ -25,28 +25,13 @@ For synchronous (i.e. non-bulk) requests, BFD users SHALL include meaningful val
 * `BlueButton-UserId`: the unique ID in the frontend system for the user (e.g. beneficiary) that data is being requested on behalf of.
 * `BlueButton-User`: the human-readable login ID (e.g. email address) in the frontend system for the user (e.g. beneficiary) that the data is being requested on behalf of.
 * `BlueButton-BeneficiaryId`: the unique beneificary ID (e.g. `Patient.id`) from BFD for the user the beneficiary whose data is being requested.
+* `x-forwarded-for`: a standard header for identifying the originating IP address of a client connecting to a web server through an HTTP proxy or a load balancer.
 
 ## Asynchronous Requests
 
 For asynchronous (i.e. bulk) requests, BFD users SHALL include meaningful values for as many as possible of the following HTTP headers.
 
-### BCDA
```
Comment (msnook) [inline docs/request-audit-headers.md:33]: Will requests that contain these deprecated headers be rejected?
For reference, the code in BCDA which sets these headers is [here](https://github.com/CMSgov/bcda-app/blob/master/bcda/client/bluebutton.go#L286-L310).
CC: @mtrang1263

File: docs/request-audit-headers.md [hunk 2]
```diff
@@ -25,28 +25,13 @@ For synchronous (i.e. non-bulk) requests, BFD users SHALL include meaningful val
...
-* `AB2D-CMSID`: the Part D contract ID
-* `AB2D-JOBID`: integer generated by DB sequence
-
-### DPC
-
 * `BlueButton-OriginalQueryId`: (see above; same thing)
     * For asynchronous jobs, users SHALL use the `BlueButton-OriginalQueryId` for the HTTP request that created/submitted/whatever the job.
-* `DPC-CMSID`: The NPI of the downstream requesting provider or organization.
-* `DPC-JOBID`: integer generated by DB sequence
+* `BULK-CLIENTID`: a unique identifier of the bulk client for whom this request is for (i.e. ACO ID, NPI, Part D Contract)
```
Comment (mtrang1263) [inline docs/request-audit-headers.md:36]: What is the recommended migration path for these headers?

Should BCDA start sending `BULK-CLIENTID` and `BULK-JOBID` in addition to `BCDA-CMSID` and `BCDA-JOBID`?

And if we do not send these two headers (and continue sending `BCDA-CMSID` and `BCDA-JOBID`), will our request get rejected?

File: docs/request-audit-headers.md (hunk 1 above)
Comment (jzulim) [reply docs/request-audit-headers.md:33]: No. These headers are only used for auditing purposes.

File: docs/request-audit-headers.md (hunk 2 above)
Comment (jzulim) [reply docs/request-audit-headers.md:36]: @mtrang1263 No soft migration is necessary. We can cut over in a hard fashion. These headers are used for audit purposes only. These specific bulk headers do not impact the validity of requests to BFD.

File: docs/request-audit-headers.md [hunk 3]
```diff
@@ -25,28 +25,13 @@ For synchronous (i.e. non-bulk) requests, BFD users SHALL include meaningful val
 * `BlueButton-UserId`: the unique ID in the frontend system for the user (e.g. beneficiary) that data is being requested on behalf of.
 * `BlueButton-User`: the human-readable login ID (e.g. email address) in the frontend system for the user (e.g. beneficiary) that the data is being requested on behalf of.
 * `BlueButton-BeneficiaryId`: the unique beneificary ID (e.g. `Patient.id`) from BFD for the user the beneficiary whose data is being requested.
+* `x-forwarded-for`: a standard header for identifying the originating IP address of a client connecting to a web server through an HTTP proxy or a load balancer.
 
 ## Asynchronous Requests
```
Comment (mtrang1263) [inline docs/request-audit-headers.md:30]: Do we need another header to differentiate between the various partner applications? Or do we believe that `BULK-CLIENTID` can be leverage to see who the source was (e.g. BCDA request would have BULK-CLIENTID set to A****, V***, E***)

Comment (jzulim) [reply docs/request-audit-headers.md:30]: We record the TLS certificate presented in the handshake as another header, which we use to key off.

**Review by mtrang1263 — DISMISSED** [review]
Approach looks good to me. In terms of timing, I think we can slot this in our upcoming sprint, but will confirm with Laureen to give a more definitive time window.

File: docs/request-audit-headers.md [hunk 4]
```diff
@@ -25,28 +25,13 @@ For synchronous (i.e. non-bulk) requests, BFD users SHALL include meaningful val
...
-* `AB2D-JOBID`: integer generated by DB sequence
-
-### DPC
-
 * `BlueButton-OriginalQueryId`: (see above; same thing)
     * For asynchronous jobs, users SHALL use the `BlueButton-OriginalQueryId` for the HTTP request that created/submitted/whatever the job.
-* `DPC-CMSID`: The NPI of the downstream requesting provider or organization.
-* `DPC-JOBID`: integer generated by DB sequence
+* `BULK-CLIENTID`: a unique identifier of the bulk client for whom this request is for (i.e. ACO ID, NPI, Part D Contract)
+* `BULK-JOBID`: integer generated by DB sequence
```
Comment (DeirdreHolub) [inline docs/request-audit-headers.md:37]: We use UUIDs for almost everything in the backend, including job ids. We do not use db sequences. Is an integer really a must?

File: docs/request-audit-headers.md (hunk 2 above)
Comment (DeirdreHolub) [inline docs/request-audit-headers.md:36]: DPC uses UUIDs as internal identifiers, as mentioned elsewhere in another comment. DPC logging always includes the UUID, not the NPI. Any problem with using UUID here?

Comment (jzulim) [reply docs/request-audit-headers.md:36]: This will be ok for us. Thanks.

File: docs/request-audit-headers.md (hunk 4 above)
Comment (jzulim) [reply docs/request-audit-headers.md:37]: This is also ok.

**Review by keithdadkins — APPROVED** [review]
LGTM!

**Comment by njdister** [comment]
I had a look around the BB2.0 code and found this header:
```
BlueButton-OriginatingIpAddress
```
Is that being picked up by BFD?  If so, should it be replaced with x-forwarded-for?

**Review by keithdadkins — APPROVED** [review]
LGTM

**Review by cbrunefearless — APPROVED** [review]
lgtm

**Pull request merged**